### DISCUSSION TITLE
case-insensitive strings

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#5545] [Design Proposals]
I think it would be interesting to have built-in support for some sort of case-insensitive string type, where:

- dirty checking would use `equalsIgnoreCase()`
- UKs in schema generation would use `lower()`
- `=` and `<` comparisons in HQL would use `lower()`
- perhaps, we would lowercase the string before sending it to JDBC.

The idea came to me here https://github.com/jakartaee/persistence/issues/209#issuecomment-1311475512.

**Comment by beikov** [comment]
I love the idea. It's useful for e.g. Email addresses and domain names which are all case insensitive. In the past I used a PostgreSQL type called `citext` for this, but I think if we have a dedicated `JavaType` for this, that would be very useful.

**Reply by gavinking** [reply]
Well, hell, if some dbs have a type for it, we could even look into supporting that in DDL generation. That would be nice.

**Reply by beikov** [reply]
Dunno if we really should emit that DDL type by default. There seem to be some caveats with the built-in type: https://www.postgresql.org/docs/current/citext.html

I guess that should be configurable. Maybe we should consider some `@Text` annotation that also allows specifying this and optionally also the collation to be used?

**Reply by gavinking** [reply]
Ah OK, if it's a doubtful thing then no, better to stick to standard SQL types. TBH I had never heard of it before.
